### ISSUE TITLE
Standardisation of Audit Log objects

### ISSUE THREAD

**Issue by joepavitt** [flowforge/flowforge#1183]
### Epic/Story
https://github.com/flowforge/flowforge/issues/1166

### Blocking/Slowing Down
#428 and #517

### Description

As part of #1084 @Steve-Mcl mentioned [here](https://github.com/flowforge/flowforge/pull/1084#issuecomment-1280701797) about standardising our audit logging structure, whilst that work was agreed to be valuable, it was not developed there and then due to scope creep from the original issue.

Currently, I am tasked with developing #428 and #517 which utilises the provided audit log from the API to display it in a user-friendly format in the UI. This is proving a very tedious if/else exercise due to the unstandardised formatting of the logs, particularly for different keys being used to represent the same entity across different log types.

This issue will scope the discussion of a new structure upon which all logs can standardise to.

**Comment by joepavitt** [comment]
My proposal would be that keys inside the `body` of a log entry are not allowed to be generic features, e.g. `id`, `name`, but should be nested in context/scope for what those features belong to, e.g. `project.id`, `project.name`. 

This ensures consistency for the use of those keys across all events as currently an `id` in the body of one event could refer to the `team.id`, but an `id` in another could be the `project.id`.

In assessing the existing logs, it would lead me to the following structure/keys available on the `body` of an audit log:

| key | description |
| --- | ------------ | 
| team | team object with at least one of the following keys: `{ id, name, type, slug }` |
| project | project object with at least one of the following keys: `{ id, name }` |
| device | device object with at least one of the following keys: `{ id, name }` |
| user | user object with at least one of the following keys: `{ name, username, email, role }` |
| stack | stack object with at least one of the following keys: `{ id, name }` |
| settings | platform settings object  |
| license | the body of a FF license, `{ key }` |
| billing_session |  the billing session id from Stripe |

I may have missed some, but think that's everything we use for now.

**Changing/Updating Values**

in any instance where we we have recorded change in values, I propose:

| key | description |
| --- | ------------ | 
| new | an object constrained to the above keys, detailing the keys that have been changed |
| old | an object constrained to the above keys, detailing the keys that have been changed |

example:
```
{
   old: {
      team: {
         name: 'Old Team Name',
         id: '1a2b3c'
      }   
   },
   new: {
      team: {
         name: 'New Team Name',
         id: '1a2b3c'
      }   
   }
}
```

**Structuring Errors**

As per @Steve-Mcl's work in #1084, my proposal would be:

```

| key | description |
| --- | ------------ | 
| code | snake_case identifier for the error type |
| error | text-based description of the error, offering more insight into what went wrong |
| ... | any of the other relevant keys from the first table above |

```

**Comment by joepavitt** [comment]
cc'ing @hardillb and @knolleary as important to get their input here too.

**Comment by Steve-Mcl** [comment]
Hi Joe, on the face of it, I completely agree the lack of standard will make it difficult to complete your tasks and make it harder the longer we leave this. So I am +1 for some work in this area (and happy to implement too if priorities permit)


---

> user	user object with at least one of the following keys: `{ name, username, email, role }`

There are instances where a user name or id is not available.  The same may be true for other objects (either now or in the future). For that reason, I would propose each of these standard objects support a `message` field for adding flavour to a log entry.  e.g. consider this pseudo entry: `{ team: { id: null, message: 'The team ID should NOT be NULL. Please report this to FlowForge'} }`  

---

I would propose we also develop and use helper functions for structuring these standard logging objects.  That way no assumptions are made about parameters and required fields. As a benefit, any improvements/fixes are done in one place only.

--- 

Should there be a `severity` or `log_level` key somewhere? 

---

**Comment by joepavitt** [comment]
> For that reason, I would propose each of these standard objects support a message field for adding flavour to a log entry. e.g. consider this pseudo entry: { team: { id: null, message: 'The team ID should NOT be NULL. Please report this to FlowForge'} }

Is that not an error though, and therefore covered by the error structure proposed?

**Comment by joepavitt** [comment]
> Should there be a severity or log_level key somewhere?

Not sure we need a severity/level beyond `log` and `error` for now?

**Comment by knolleary** [comment]
@joepavitt I'm happy with your proposal as it stands. I suspect there will be some edge cases that don't quite fit, but we should deal with them if/when they come up.

> For that reason, I would propose each of these standard objects support a message field for adding flavour to a log entry. e.g. consider this pseudo entry: { team: { id: null, message: 'The team ID should NOT be NULL. Please report this to FlowForge'} }

Agreed with @joepavitt - that would be an error so should be formatted as an error body.

I don't think we need severity or log_level at this stage. We can revisit that if it becomes clear there is a need for more granularity around log events and error events.

**Comment by Pezmc** [comment]
The proposed spec above "with at least one of the following keys" reads as if a valid record would just be `{ name }` or `{ type }`. 

I think it should rather be "with at least the following keys". We need to store enough information to display sensible audit logs, otherwise future changes will affect past log entries (if we're looking up the latest copy of the object for info).

e.g. if a project, for example, changed name would you expect:

 - Project 1 was edited 
 - Project 1 was renamed to Project 2
 - Project 2 was deleted

Or

- Project 2 was edited (latest copy looked up using ID)
- Project 1 was renamed to Project 2
- Project 2 was deleted
 
-----

> I would propose we also develop and use helper functions for structuring these standard logging objects. That way no assumptions are made about parameters and required fields. As a benefit, any improvements/fixes are done in one place only.

Strong 👍 - even something as specific as `AuditLog.logCreateProject(project)`, `AuditLog.logDeleteProject(project)`, `AuditLog.logUpdateDevice(oldState, newState)`. While it might feel overly verbose now, it'll make our lives easier in the long run and ensure standardisation.

-----

Following the KISS principle, having errors only at the body level seems a good start. It can always be extended to the individual object level later, if a specific need arrises.

**Comment by joepavitt** [comment]
Agree @Pezmc - the body of an object (team, device, project) should contain as many of the listed keys as possible. My plan would then be to build the messaging client-side, and the audit logging server-side remains object key/value based only. This gives us flexibility on messaging as cases evolve, or we want to make minor changes to language, but maintain consistency across all historical logs of the same type.

**Comment by Steve-Mcl** [comment]
@joepavitt 
> > For that reason, I would propose each of these standard objects support a message field for adding flavour to a log entry. e.g. consider this pseudo entry: { team: { id: null, message: 'The team ID should NOT be NULL. Please report this to FlowForge'} }

> Is that not an error though, and therefore covered by the error structure proposed?


@knolleary 
> Agreed with @joepavitt - that would be an error so should be formatted as an error body.



That particular one was yes.

However, consider an action that is possible via 2 or 3 events/routes. We might see `{ project: { id: 'pr1'} }` - we know it was suspended or started or stopped (because of the `event` field) but was it a user action? an auto action? an anomaly? So my suggestion was that with a little flavour it becomes more apparent e.g. `{ project: { id: 'pr1'}, message: 'Project Started by user' }` 

It is certainly a nice to have and would/could simplify future diagnostics for both us and end users so if we do not add the ability to decorate the logs in this iteration that is fine so long as this work doesnt block any future additions, then I am happy to put that aside 👍

**Comment by Steve-Mcl** [comment]
@Pezmc +1 on "with at least the following keys"

> I think it should rather be "with at least the following keys". We need to store enough information to display sensible audit logs, otherwise future changes will affect past log entries (if we're looking up the latest copy of the object for info).

**Comment by joepavitt** [comment]
> However, consider an action that is possible via 2 or 3 events/routes. We might see { project: { id: 'pr1'} } - we know it was suspended or started or stopped (because of the event field) but was it a user action? an auto action? an anomaly? So my suggestion was that with a little flavour it becomes more apparent e.g. { project: { id: 'pr1'}, message: 'Project Started by user' }

I would actually expect, and I now realise I missed this from my original proposal, that _all_ events have an "author", so we know which user has triggered it.

**Comment by joepavitt** [comment]
Finally, also now realising the top level `id` of an event log is actually just the id of what I would call the "scope". So if it's a team-event, then the `id` of the event log is the id of the team...

Feel like we should have clarity here too. Maybe a `scope` key, which then a `type` and `id` key embedded in that object?

**Comment by knolleary** [comment]
For each log entry there needs to be:
 - what was operated on
 - who did it
 - what happened

In the database table, each entry has the fields `entityId`, `entityType` and `body`.

The `entity*` fields identify the scope of the event - the team/project/user or platform-wide.

If we take the example of 'user invited', that is a team event - so the entity fields will identify the team. I would suggest also having `team` in the body would be redundant. For that particular case, the body needs to identify both *who* did it and *who* was invited. So should `user` be the person doing the thing, or the person being acted on. We should standardise that.

> However, consider an action that is possible via 2 or 3 events/routes. We might see { project: { id: 'pr1'} } - we know it was suspended or started or stopped (because of the event field) but was it a user action? an auto action? an anomaly?

If a project is stopped, then the audit log entry needs to identify why. If it was stopped due to a user request, then the audit log entry should include a `user` property identifying that user. If it was stopped for any other reason, that needs to be included.

**Comment by Steve-Mcl** [comment]
> > However, consider an action that is possible via 2 or 3 events/routes. We might see { project: { id: 'pr1'} } - we know it was suspended or started or stopped (because of the event field) but was it a user action? an auto action? an anomaly?
> 
> If a project is stopped, then the audit log entry needs to identify why. If it was stopped due to a user request, then the audit log entry should include a `user` property identifying that user. If it was stopped for any other reason, that needs to be included.

That is indeed a better approach but I suspect we will have to (or aim to) do a retro on the existing `event`s to ensure this is the case?



> If we take the example of 'user invited', that is a team event - so the entity fields will identify the team. I would suggest also having `team` in the body would be redundant. For that particular case, the body needs to identify both _who_ did it and _who_ was invited. So should `user` be the person doing the thing, or the person being acted on. We should standardise that.

Very good point.  I updated the GSheet in an attempt to clarify this as it is not immediately obvious. Having this information encoded in the log entry in a formalised manner would definitely be a +1 from me.

**Comment by joepavitt** [comment]
okay, given the above, I put forward the following:

```
{
   "createdAt": <datetime>,
   "event": <event>,
   "scope": {
      "id": <entity-id>,
      "type": <team, device, platform, etc.>
   },
   "trigger": {
      <user details if a user>
      ... not dsure what alternatives are here, open to suggestions   
   }
   "body": <as per my descriptions above>   
}
```

I _believe_ the _"who did it"_ is currently nested in the `body`?

**Comment by knolleary** [comment]
@joepavitt that makes sense in terms of the format returned in by the audit log api. In terms of how that would map to the database table, we have `entityId` and `entityType`. Changing those to be nested properties under a `scope` column would make the table hard to index efficiently. So this would have to be done as a transformation between the database and the API response.

For the `trigger` - we would either have to add `triggerId` and `triggerType` to the database table, or store a trigger object under `body` which gets moved to the top level as part of the xform between database and API response. Either approach has pros and cons

**Comment by joepavitt** [comment]
I wasn't necessarily proposing any changes to the DB table, `entityId` and `entityType` are find in the DB - transformation is cleaner imo.

To be consistent, I would add `triggerId` and `triggerType` and transform as per `entity` > `scope`

**Comment by joepavitt** [comment]
DB structure has one job, API structure has another - I am very anti making them identical - they each serve a different purpose, and thought should be put into both on how we effectively communicate the data at the DB/API levels

**Comment by joepavitt** [comment]
Proposed steps (agreed on call with Steve):

- Return existing DB content in API format proposes [above](https://github.com/flowforge/flowforge/issues/1183#issuecomment-1303272804)
   - For now, and for handling of historical data, `trigger` can be empty
   - For stage 1 - no changes to `body`
- Update log stores to include `triggerId` and `triggerType` columns for all future audit logs.
- Create a helper function/class for triggering/recording audit logs in format TBD
   - Update all uses/triggering of audit logs to use this function
   - Include `trigger...` for each new entry
- Reformat the `body` as per https://github.com/flowforge/flowforge/issues/1183#issuecomment-1302298350

**Comment by Steve-Mcl** [comment]
Following discussion with Joe, the below points were ratified...
1. `event` strings should ALWAYS be present tense (e.g. `project.settings.update` and `platform.licence.inspect`, `project.create`) 
    1. This means an action is still logged even if it fails (failures should be caught and logged with the `code` and `error` as per previous comment)
    2. This improves readability like `account.login` where an attempt is made but not successful
1. The `trigger` object can be extrapolated from the user (or system app) that actioned the op.
    1. User -1 will be used to signify an "anonymous user"
    2. User 0 will be used to signify a "system" operation
    3. User `null` is accepted for historical entries (defaults to "Unknown")
    4. a `name` field should be included and populated accordingly (for simplified UI display)
1. If a log entry contains `code` or `error` these will be delivered to the UI API in the property `error` with the following properties at minimum: `{code: "camel_case_code", message: "Error Text"}` 
1.  `new` and `old` objects should be in the body under `updates` e.g. `body.updates.new.name = "new name"` and `body.updates.old.name = "old name"`
1. Long handed code that abstracts the logging to DB with well defined function signatures ensuring the correct items are passed and that the correct elements are recorded every time. Example: `teamAuditLog.user.added` has a function signature of `async added (actionedBy, team, user)` meaning it MUST be passed a `team` and `user` object thus ensuring the logged parts are present and uniform
1. All body objects (into the database) are to be passed though a set of formatter helpers to ensure data is uniformly formatted to specification

**Comment by Steve-Mcl** [comment]
@joepavitt @Pezmc 

With the attached PRs merged, https://github.com/flowforge/flowforge/issues/428 and https://github.com/flowforge/flowforge/issues/517 will be unblocked.

There is some additional work required to update the GSheet to list out the `event` names - Given time,  I would of course pick this up - BUT - with all log entries now in one place it is very easy to determine what they are & what additional objects they can have - potentially making the upkeep of a separate document redundant? As an example...
The [project.deleted](https://github.com/flowforge/flowforge/blob/ef24b48a636a664a08deea122b8db85d6278099f/forge/lib/audit-logging/project.js#L13) `event` will have body objects `error`, `team` & `project`  (error is excluded when the operation is successful). 
The format of the `body.project` is clearly spelled out in the [`projectObject`](https://github.com/flowforge/flowforge/blob/e133d460ec604d1535693a8b99a35b1a1a61443b/forge/lib/audit-logging/formatters.js#L145) helper function ...

```
{
        id: project?.id || null,
        hashid: project?.hashid || null,
        name: project?.name || unknownValue
}
```
NOTE: `hashid` is only included for DB reasons - your UI can exclude any parts of the object it needs to.


Lastly, in the end, I did NOT rename the **tense** of all `event`s (due to time) nor did I (as much as I really wanted to) rename the `events` to a common theme. I feel the correct course of action should be to standardise the naming of the `event`s (in both components and tense) as there are currently differing strategies.  For example, look at this selection of event names...
* 'team.user.roleChanged'
* 'team.device.credentialsGenerated'
* 'account.forgot-password'
* 'users.auto-create-team'
* 'project.snapshot.deviceTarget' 

The first 2 are clearly `team` scope but use camelCase. The next 2 use snake-case and are part of the `user` scope and but that is not immediately clear without accompanying data.  As for the last entry - what does that even mean? 

---

For now, I simply must press on with the file nodes work - please let me know if you get (or dont get) the GSheet updated (or if it is even worth maintaining now due to the centralised location of audit logging.

Ta.

**Comment by joepavitt** [comment]
Will be making them consistently past tense as it reads better when conducting the logging itself, otherwise it feels like `projectLogger.create()` for example is running that action, `projectLogger.created()` I think shows a stronger link to what that function call is actually doing, in that it's recording the event.

**Comment by Steve-Mcl** [comment]
> Will be making them consistently past tense as it reads better when conducting the logging itself, otherwise it feels like `projectLogger.create()` for example is running that action, `projectLogger.created()` I think shows a stronger link to what that function call is actually doing, in that it's recording the event.

Completely fine by me Joe. Consistency is definitely more important.

**Comment by knolleary** [comment]
I'm doing a bit of rework on #1224 to help make it testable. Whilst going through it I have spotted this comment above:

>  NOTE: hashid is only included for DB reasons - your UI can exclude any parts of the object it needs to.

As per our [api doc guidelines](https://github.com/flowforge/flowforge/blob/main/docs/contribute/api-design.md#object-ids) we should *never* return the raw database `id` property on the external API. We use the value of `hashid` in place of `id`.

As part of the work I'm doing, I'll modify `formatLogEntry` in order to sanitize the output it returns, but I wanted to highlight this as it's an important part of our API design.

**Comment by ZJvandeWeg** [comment]
Moving this to 1.2 as I understand the target merge date is next week Monday? @Steve-Mcl Please also update the documentation, I've not spotted those changes yet.

**Comment by Steve-Mcl** [comment]
A random test (approx 40%) test on staging + 100% test on local.

All sub functions (where the actual logging occurs for Team/Project/User/Platform events to the 3 scopes Platform/Team/Project) have been executed

**Comment by joepavitt** [comment]
> There is some additional work required to update the GSheet to list out the event names 

this has now been updated to align with all events provided by the new logger
